src/utils/narrow.js

```diff
@@ -282,11 +289,13 @@ export const parseNarrow = (narrowStr: string): Narrow => {
     }
 
     case 'pm:': {
-      if (!rest.startsWith('s:')) {
+      const match = /^d:(.*?):(.*)/.exec(rest);
```

Comment:
Ah OK, I think I've worked out what we're doing here:

The `?` in the first capture group, `(.*?)`, makes the `*` quantifier lazy instead of greedy, which it would be by default. This is fine because the IDs are numeric and obviously don't contain a `:`. It's also necessary, because email addresses might contain colons (er, I think? [Wikipedia](https://en.wikipedia.org/wiki/Email_address#Domain) gives the example "jsmith@[IPv6:2001:db8::1]", which I think may be permitted by [IETF RFC 2821 section 4.1.3](https://tools.ietf.org/html/rfc2821#section-4.1.3)).

It'll be simpler to think about once we've taken emails out of the picture. If that weren't planned, I'd say we might want to have an explanatory code comment here.

This reminds me of what we're doing for topic narrows, starting in af1713a76 (though it's seeming like that'll need to be reworked because it caused a regression; the `\x00` character we use there (which has a very similar role to the `:` here) doesn't round-trip into being stored as the value of the `data-narrow` attribute in our message-list HTML; see [discussion](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.23M4339.20regression/near/1085234)).

Reply:
> The `?` in the first capture group, `(.*?)`, makes the `*` quantifier lazy instead of greedy, which it would be by default.

Yep!


> It's also necessary, because email addresses might contain colons (er, I think? [Wikipedia](https://en.wikipedia.org/wiki/Email_address#Domain) gives the example "jsmith@[IPv6:2001:db8::1]", which I think may be permitted by [IETF RFC 2821 section 4.1.3](https://tools.ietf.org/html/rfc2821#section-4.1.3)).

Yeah, good find with that example. I didn't even think about whether an email address could contain a colon. But the great thing about using `.*?` here is that we don't have to, because it will match up to that first colon (which has to be the one we put there as a delimiter), and then everything (*) after that is the email addresses.

In general for capturing contents up to a delimiter, `.*?` is what you want rather than `.*`, for just this reason. It similarly appears in the other regexps in this `parseNarrow` function.

I did deliberately put the IDs first and the emails at the end, so that the emails would be the part in that position where we don't have to think about what's there.

(*) Though looking again at the `topic:` case reminds me that there is an asterisk on that, because without the `/s` modifier there are a few characters the `.` pattern rejects. I'm not sure if those characters are possible in email addresses -- I'm sure ASCII newline isn't, but there are a couple of other "newline" characters that wrinkle applies to and I'm not certain if they can appear. I'll add `/s` there too so we don't have to think about it.


> (though it's seeming like that'll need to be reworked because it caused a regression; the `\x00` character we use there (which has a very similar role to the `:` here) doesn't round-trip into being stored as the value of the `data-narrow` attribute in our message-list HTML; see [discussion](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.23M4339.20regression/near/1085234)).

(We ended up fixing that issue at a different layer, in #4352.)